runtime/interpreter/interpreter_expression.go

```diff
@@ -819,10 +819,23 @@ func (interpreter *Interpreter) VisitReferenceExpression(referenceExpression *as
 
 	result := interpreter.evalExpression(referenceExpression.Expression)
 
-	return &EphemeralReferenceValue{
-		Authorized:   borrowType.Authorized,
-		Value:        result,
-		BorrowedType: borrowType.Type,
+	switch result := result.(type) {
```

Comment:
I'm not quite sure, but I think the type check shouldn't be on the dynamic type of the value, but the static type of the referenced expression

Reply:
We do need to do this switch on the dynamic type so we know whether or not to move the `*SomeValue` constructor. However you are right in that this check should only occur if the static type of the expression is itself optional.